README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![License:
MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/parlitools)](https://cran.r-project.org/package=parlitools)
[![GitHub
tag](https://img.shields.io/github/tag/evanodell/parlitools.svg)](https://github.com/evanodell/parlitools)
[![](https://cranlogs.r-pkg.org/badges/grand-total/parlitools)](https://dgrtwo.shinyapps.io/cranview/)
[![Travis-CI Build
Status](https://travis-ci.org/evanodell/parlitools.svg?branch=master)](https://travis-ci.org/evanodell/parlitools)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/evanodell/parlitools?branch=master&svg=true)](https://ci.appveyor.com/project/evanodell/parlitools)
[![Coverage
Status](https://img.shields.io/codecov/c/github/evanodell/parlitools/master.svg)](https://codecov.io/github/evanodell/parlitools?branch=master)
[![DOI](https://zenodo.org/badge/86801920.svg)](https://zenodo.org/badge/latestdoi/86801920)

# parlitools

A collection of useful tools for UK politics, including base maps and
datasets. Initially inspired by Bhaskar Karambelkar’s
[`tilegrams`](https://cran.r-project.org/package=tilegramsR) package,
but with the ability to create a hexagonal map of UK parliamentary
constituencies and local authorities. The package also includes
functions for data retrieval of current MPs and their consituency
details (as it requires calls to two different APIs, this function is
not included in my
[`hansard`](https://cran.r-project.org/package=hansard) or
[`mnis`](https://cran.r-project.org/package=mnis) packages), and local
government composition. There are inbuilt datasets from the 2015 and
2017 UK General Elections and 2011 Census, courtesy of the British
Election Study, estimates of EU referendum votes from Chris Hanretty,
and hex codes for different party colours.

The most up-to-date documentation will always be at
<http://docs.evanodell.com/parlitools/>.

## Installing

`parlitools` is available on CRAN. To install on CRAN run:

    install.packages("parlitools")

Or, if you use [pacman](https://cran.r-project.org/package=pacman):

    pacman::p_load(parlitools)

To install the development version from GitHub, run:

    ##install.packages("parlitools")
    devtools::install_github("evanodell/parlitools")

## Functions and Data

For more details see [the full
documentation](https://docs.evanodell.com/parlitools/reference) or the
[vignettes](https://docs.evanodell.com/parlitools/articles/).

### Included Data

`party_colour` - A tibble with the ID, name and hex code for the
official colour of a variety of political parties, taken from Wikipedia.
Includes all political parties with MPs and a number without MPs.
(Sources:
<https://en.wikipedia.org/wiki/Wikipedia:Index_of_United_Kingdom_political_parties_meta_attributes>,
[`mnis::ref_parties()`](https://cran.r-project.org/package=mnis))

`bes_2015` - A tibble with the British Election Study 2015 Constituency
Results Version 2.2. For information on all the variables in this
dataset, see the [bes-2015
vignette](http://docs.evanodell.com/parlitools/articles/bes-2015.html)
(Source:
<http://www.britishelectionstudy.com/data-object/2015-bes-constituency-results-with-census-and-candidate-data/>)

`bes_2017` - A tibble with Great Britain constituencies results from the
2017 general election. This data can be linked to 2011 census
information in
[`census_11`](https://docs.evanodell.com/parlitools/reference/bes.html).

`census_11` - A tibble with constituency-level census data.

`leave_votes_west` - The percentage of votes cast for leave in the 2016
EU referendum. Some constituencies have actual results and others only
have estimates by Chris Hanretty; in cases where the actual cote count
is known, both the estimates and the actual results are reported.
(Sources: Hanretty, C. (2017). Areal interpolation and the UK’s
referendum on EU membership. *Journal of Elections, Public Opinion and
Parties*, 27(4), 466–483.
<https://doi.org/10.1080/17457289.2017.1287081>,
<http://www.bbc.co.uk/news/uk-northern-ireland-36616830>)

### Data Retrieval Functions

`current_mps` - Uses functions from `hansard` and `mnis` to create a
tibble with data on all current MPs, their party affiliation and their
constituency.

`mps_on_date` - Uses functions from `hansard` and `mnis` to create a
tibble with data on all MPs from a given date, their party affiliation
and their constituency.

`west_hex_map` - A hexagonal cartogram, stored as a simple feature and
data frame, of Westminster parliamentary constituencies. `west_hex_map`
can be used to create maps like this:

<img src="tools/hex_map.png" width="456px" />

`local_hex_map` - Hexagonal cartogram, A hexagonal cartogram, stored as
a simple feature and data frame, of all Local Authorities in England,
Wales and Scotland.

### Using `parlitools`

For more details, please see the [introductory
vignette](https://docs.evanodell.com/parlitools/articles/introduction.html),
[using `parlitools` with
`cartogram`](https://docs.evanodell.com/parlitools/articles/using-cartograms.html),
[mapping local
authorities](https://docs.evanodell.com/parlitools/articles/mapping-local-authorities.html)
and the vignette detailing [British Election
Study 2015](https://docs.evanodell.com/parlitools/articles/bes-2015.html)
variables.

## Data Sources

There are a variety of potentially relevant data sources and datasets on
UK politics, far too many for me to include them all in this package.

  - [Electoral
    Commission](http://www.electoralcommission.org.uk/our-work/our-research/electoral-data)
    - Electoral results dating back to 2005.

  - [British Election Study](http://www.britishelectionstudy.com/data/)
    - A large selection of open data, including panel surveys, linked
    data and aggregated Twitter data, covering elections and referenda.

  - My [`hansard`](https://cran.r-project.org/package=hansard) &
    [`mnis`](https://cran.r-project.org/package=mnis) data retrieval
    packages for parliamentary APIs.

  - [Open Council Data](http://opencouncildata.co.uk/) has data on the
    names, parties, and wards of all UK councillors, updated more or
    less weekly. `parlitools` uses this site to power the
    `council_seats()` function.


NEWS.md


# parlitools 0.4.1.9000

* `mps_on_date()` no longer imports redundant constituency data.

* updates to `party_colours` dataset

# parlitools 0.4.1

* Fixed some missing 2017 results from `bes_2019` data, spotted by Antony Unwin.

# parlitools 0.4.0

* Added 2019 UK General Election Results

* Minor vignette improvements

# parlitools 0.3.4

* Fixed issues being caused by some old vignettes.

# parlitools 0.3.3

* Fixed broken links in README

# parlitools 0.3.2

* Changed `gss_code` to `ons_const_id` in the `leave_votes_west` dataset
  to ensure consistent naming of parliamentary constituency codes
  
* Added Northern Ireland GE results


# parlitools 0.3.1

* Fixed formatting of percentages in the `leave_votes_west` dataset so they are 
  all scaled on 0-1 rather than some on 0-1 and some on 0-100.
  
* Improved the organisation of the British Election Study dataset documentation.


# parlitools 0.3.0

* `parlitools` no longer stores local council party data. The most up-to-date
  data on the political make-up of local councils and individual councillors
  can be be accessed with the new `council_seats()` function.

* Updated the `leave_votes_west` dataset to include the latest best-estimates
  of constituency level votes from Chris Hanretty and the actual votes, where
  known.

* Fixed errors caused by duplicated vignette names.

* Eliminated some dependencies.

# parlitools 0.2.3

* Changed `party_id` in the `party_colour` dataset to a character vector.

* Otherwise very minor update, some tweaks to docs layout.

# parlitools 0.2.2

## Data

* Updating of `council_data` dataset to be accurate as of 2017-11-24.

# parlitools 0.2.1

## Data

* Updating of `council_data` dataset to be accurate as of 2017-09-13.

* Added `combined_leave_vote` column to `leave_votes_west` dataset with the 
actual leave vote if known and the Chris Hanretty estimate if not known. 

* Vote percentages in `leave_votes_west` are now all numeric class.

## Dependency changes

* Removed reliance on now-hidden functions from `mnis` and `hansard` packages.

## General improvements

* Improvements to documentation to improve clarity and readability. 

* Better example plots in introductory vignette.

* The tidying functions are now all internal to the package and not exported.

# parlitools 0.2.0

## Data

* New `bes_2017` dataset, with results of 2017 General Election, taken from 
  the 2017 British Election Study. This dataset replaces the `ge_2017` dataset.
  It also replicates some of the data in the `bes_2015` dataset.

* Census data from the British Election Study has been moved to its own 
  separate file, named `census11`.

## Fixes

* Naming conventions for variables now places an underscore between each 
  distinct word (eg `onsconst_id` is now `ons_const_id`).

# parlitools 0.1.0

* Changes to vignettes to comply with CRAN policies and to reduce the time 
  needed to check the package.

* Reduction in the number of similar vignette examples to reduce build time.

## Data

* Updating of `council_data` dataset to be accurate as of 2017-07-14.

* New `ge_2017` dataset, with results of 2017 General Election, linked with 
  constituency data from `bes_2015`.

## Bug fixes

* Consistent naming of vote variables in `bes_2015`.

# parlitools 0.0.7

* Updated documentation, more comprehensive testing. Added Northern Irish EU 
  Referendum results to the `leave_votes_west` dataset.

# parlitools 0.0.6

## Bug fixes

* Fixed bugs in `mps_on_date()` that were causing problems on Windows.

# parlitools 0.0.5

## Data updates

* Added `leave_votes_west` containing estimates and - where available - official 
  counts of the proportion of leave votes in the 2016 EU referendum.

# parlitools 0.0.4

## Data updates

* Updated `bes_2015` to use V2.21 from the British Election Study.

* Added `council_data` dataset, with details on the parties controlling local 
  councils across the UK.

## Bug Fixes

* A few minor tweaks to internal code, and to the vignette examples.

# parlitools 0.0.3

## Bug Fixes

* Fixed bug in `mps_on_date` where function did not return constituency data. 
  `mps_on_date` also now accepts dates in any format or class that can be 
  converted with `as.Date()` to a date format.

* Fixed names party names in `party_colour` to align with names used in
  `bes_2015`.

* Fixed spelling mistakes in Endiburgh constituency names in `bes_2015`.

## New Features

* Added optional `tidy_style` parameter in `current_mps()` and `mps_on_date()`, 
  allowing users to decide which style snake_case, camelCase and period.case 
  they want variable names to be in, if `'tidy'==TRUE`.

# parlitools 0.0.2

* Added an additional vignette on using the data in `parlitools` to create 
  scaled maps with the `cartogram` package.

### Maps

* Now includes `local_hex_map`, a hexagonal map of all local authority 
  areas in Scotland, Wales and England.

### Data Retrieval

* The `mps_on_date` function downloads information on every MP and their 
  constituency who was a member of the house on a single given date, or 
  between two given dates.

# parlitools 0.0.1

## Introducing parlitools

* Introducing `parlitools`, a package for retrieving and analysing UK political 
  data. The package includes datasets from the British Election Study, a 
  hexagonal map of all UK constituencies, and party colour hex codes.

### Data

* The package includes a dataset with 2015 General Election results and census 
  information for each constituency in Scotland, Wales and England, taken from 
  the British Election Study.

* The package also includes a dataset (`party_colours`) with the hex code for 
  the official colours of UK political parties.

### Maps

* Includes a hexagonal map (`west_hex_map`) of all 650 constituencies in the UK.

### Data Retrieval

* The `current_mps` function downloads information on every sitting MP and 
  their constituency.


R/data.R


# west_hex_map ------------------------------------------------------------
#' A hexagonal cartogram of Westminster parliamentary constituencies,
#' stored as a simple feature.
#'
#' Originally created by Ben Flanagan at ESRI, with coordinates in Easting
#' and Northing. The coordinates in the original shapefile were converted
#' into Latitude and Longitude in QGIS, and the hexagon representing Ynys
#' Mon was moved to be closer to the mainland of Wales.
#'
#' @format An sf and data.frame object, containing 650 elements and 8 variables.
#' \describe{
#'   \item{\code{object_id}}{Row Name}
#'   \item{\code{constituency_name}}{The name of the constituency}
#'   \item{\code{description}}{Description of constituency type}
#'   \item{\code{gss_code}}{The constituency GSS (Government Statistical
#'   Service) code}
#'   \item{\code{region_name}}{Name of region within the UK}
#'   \item{\code{shape_length}}{The lengths of the shape}
#'   \item{\code{shape_area}}{The area of shape}
#'   \item{\code{geometry}}{Shape geometry/projection}
#' }
#' @source \url{http://www.arcgis.com/home/item.html?id=15baaa6fecd54aa4b7250780b6534682}
#' @rdname maps
"west_hex_map"

# local_hex_map ------------------------------------------------------------
#' Hexagonal cartogram, of all Local Authorities in England, Wales and Scotland.
#'
#' Originally created by Ben Flanagan at ESRI, with coordinates in Easting and
#' Northing. Converted coordinates into Latitude and Longitude in QGIS.
#'
#' @format An sf and data.frame object, containing 650 elements and 8 variables.
#' \describe{
#'   \item{\code{object_id}}{Row Name}
#'   \item{\code{la_code}}{Local Authority identifier code}
#'   \item{\code{la_name}}{Local Authority Name}
#'   \item{\code{shape_length}}{The lengths of the shape}
#'   \item{\code{shape_area}}{The area of the shape}
#'   \item{\code{geometry}}{Shape geometry/projection}
#' }
#' @source \url{http://www.arcgis.com/home/item.html?id=593037bc399e460bb7c6c631ceff67b4}
#' @rdname maps
"local_hex_map"

# party_colour ------------------------------------------------------------
#' A tibble with the ID, name and hex code for the official colour of a
#' variety of political parties.
#'
#' Includes all political parties with MPs and a number without MPs. It
#' also contains a 'No Overall Control' party for local government. Hex
#' codes are taken from Wikipedia.
#'
#' @format A tibble, containing 47 rows and 3 columns.
#' \describe{
#' \item{\code{party_id}}{The party ID, as assigned by the Members Name
#' Information Service}
#' \item{\code{party_name}}{The name of the political party}
#' \item{\code{party_colour}}{A hex code for the party colour,
#' as used by Wikipedia}
#' }
#' @source \url{https://en.wikipedia.org/wiki/Wikipedia:Index_of_United_Kingdom_political_parties_meta_attributes}
#' \url{http://data.parliament.uk/membersdataplatform/services/mnis/ReferenceData/Parties/}
"party_colour"


# bes_2015 ------------------------------------------------------------
#' 2015 General Election Results
#'
#' A tibble with the British Election Study
#' 2015 Constituency Results Version 2.21.
#'
#' A tibble with constituencies results from the British Election Study 2015.
#' Census data from the BES has been moved to the \code{\link{census_11}}
#' dataset, and can be linked through the with the \code{pano},
#' \code{ons_const_id} and \code{constituency_name} variables. Only contains
#' detailed results from Great Britain; the only data included on Northern
#' Irish constituencies is the party that won the seat.
#'
#' Variable names have been converted to snake_case and variables have been
#' converted to appropriate R classes. If using in publications, please
#' cite as below:
#'
#' @references Fieldhouse, Edward, Jane Green, Geoffrey Evans,
#' Hermann Schmitt, Cees van der Eijk, Jonathan Mellon, and
#' Christopher Prosser. "British Election Study, 2015: General Election
#' Results Dataset," 2015. doi:10.13140/RG.2.1.1162.1844.
#'
#' @format A tibble, containing 650 rows and 92 columns. For full details on 
#' the included variables see the vignette:
#' <http://docs.evanodell.com/parlitools/articles/bes-2015.html>
#' @source \url{http://www.britishelectionstudy.com/data-object/2015-bes-constituency-results-with-census-and-candidate-data/}
#' @seealso bes_2017
#' @seealso bes_2019
#' @seealso census_2011
"bes_2015"


# bes_2017 ------------------------------------------------------------
#' 2017 General Election Results
#'
#' A tibble with the British Election Study 2017
#' Constituency Results Version 1.0.
#'
#' A tibble with Great Britain constituencies results from the
#' 2017 General Election. This data can be linked to 2011 census
#' information in \code{\link{census_11}} with the \code{pano},
#' \code{ons_const_id} and \code{constituency_name} variables. There are 
#'
#' Variable names have been converted to snake_case and variables have been
#' converted to appropriate R classes. If using in publications, please
#' cite as below:
#'
#' @references Fieldhouse, E., J. Green., G. Evans., H. Schmitt,
#' C. van der Eijk, J. Mellon & C. Prosser (2017) British Election Study
#' 2017 Constituency Results file, version 1.0.
#'
#' @format A tibble, containing 650 rows and 123 columns. For full details on 
#' the included variables see the vignette:
#' <http://docs.evanodell.com/parlitools/articles/bes-2017.html>
#' @source \url{http://www.britishelectionstudy.com/data-object/2017-bes-constituency-results-with-census-and-candidate-data/}
#' @seealso bes_2015
#' @seealso bes_2019
#' @seealso census_2011
"bes_2017"


# bes_2019 ------------------------------------------------------------
#' 2019 General Election Results
#'
#' A tibble with the 2019 UK General Election results, taken from the House
#' of Commons library and the Democracy Club crowd-sourced data on candidates.
#'
#' A tibble with Great Britain constituencies results from the
#' 2019 General Election. This data can be linked to 2011 census
#' information in \code{\link{census_11}} with the \code{pano},
#' \code{ons_const_id} and \code{constituency_name} variables.
#' 
#' Although not taken from the official British Election Study I have preserved
#' the `bes_{election}` naming convention. 
#' Data from Northern Ireland is also included in this object.
#'
#' @format A tibble, containing 650 rows and 98 columns. For full details on 
#' the included variables see the vignette:
#' <http://docs.evanodell.com/parlitools/articles/bes-2019.html>
#' @source \url{https://candidates.democracyclub.org.uk/api/docs/csv/#past}
#' @source \url{https://researchbriefings.parliament.uk/ResearchBriefing/Summary/CBP-8749}
#' @seealso bes_2015
#' @seealso bes_2017
#' @seealso census_2011
"bes_2019"

# census_11 ------------------------------------------------------------
#' Constituency demographic information
#'
#' A tibble with demographic information on each constituency, taken from
#' the 2011 census data included in the British Election Study. Can be
#' linked to the \code{\link{bes_2017}} and \code{\link{bes_2015}} datasets
#' using the \code{pano}, \code{ons_const_id} and
#' \code{constituency_name} variables.
#' 
#' @references Fieldhouse, Edward, Jane Green, Geoffrey Evans,
#' Hermann Schmitt, Cees van der Eijk, Jonathan Mellon, and
#' Christopher Prosser. "British Election Study, 2015: General Election
#' Results Dataset," 2015. doi:10.13140/RG.2.1.1162.1844.
#'
#' @format A tibble, containing 632 rows and 191 columns. For full details on 
#' the included variables see the vignette:
#' <http://docs.evanodell.com/parlitools/articles/census-11.html>
#'
#' @source \url{http://www.britishelectionstudy.com/data-object/2017-bes-constituency-results-with-census-and-candidate-data/}
#' @seealso bes_2015
#' @seealso bes_2017
#' @seealso bes_2019
"census_11"




# ni_ge_2017 --------------------------------------------------------------
#' Northern Ireland GE Results
#' 
#' A tibble with results from UK General Elections in Northern Ireland
#' in 2015 and 2017.
#' 
#' Data is taken from the Electoral Commission results. 
#' 
#' @format A tibble, with 18 rows and 62 columns.
#' 
#' @source \url{https://www.electoralcommission.org.uk/our-work/our-research/electoral-data/electoral-data-files-and-reports}
#'
"ni_ge_2017"



# la_codes ------------------------------------------------------------
#' A tibble with the name, ONS code and local authority type for UK
#' local authorities.
#'
#' @format A tibble, with 418 rows and 3 columns.
#' \describe{
#' \item{\code{name}}{Local Authority Name}
#' \item{\code{la_code}}{The Local Authority code from the Office for
#' National Statistics}
#' \item{\code{type}}{The type of local authority}
#' }
#' @source \url{http://opencouncildata.co.uk/councils.php?model=}
#' \url{http://geoportal.statistics.gov.uk/datasets/464be6191a434a91a5fa2f52c7433333_0}
"la_codes"

# leave_votes_west ------------------------------------------------------------
#' A tibble with leave votes from the Brexit referendum in each constituency.
#'
#' A tibble with details on percentage of votes cast for leave in the 2016
#' referendum on British membership of the European Union in each
#' constituency. Numbers for Britian were compiled by the House of Commons
#' using data from Chris Hanretty's estimates of leave vote, and supplemented
#' with data from BBC Freedom of Information requests for more finely grained
#' voting data. Numbers for Northern Ireland, which reported results by
#' constituency, are taken from the BBC.
#' @format A tibble, with 650 rows and 8 columns.
#' \describe{
#' \item{\code{ons_const_id}}{ONS code for constituency}
#' \item{\code{constituency_name}}{The name of the constituency.}
#' \item{\code{party_2016}}{The party holding the constituency on 23 June 2016,
#' the date of the referendum.}
#' \item{\code{ch_leave_vote}}{Estimates of the leave vote produced by
#' Chris Hanretty.}
#' \item{\code{known_leave_vote_perc}}{The percentage of votes cast for leave,
#' if known.}
#' \item{\code{how_do_we_know}}{A character vector indicating the source of 
#' known constituency level Brexit votes. "Rosenbaum / Greenwood" refers to 
#' \href{https://www.bbc.co.uk/news/uk-politics-38762034}{BBC reporting}.}
#' \item{\code{figure_to_use}}{Actual leave vote, where known, and 
#' \code{ch_leave_vote} otherwise.}
#' #' \item{\code{known_leave_vote}}{A dummy variable indicating 'Yes' if the
#' actual leave vote percentage is known, and 'No' if only the estimate by
#' Chris Hanretty is available.}
#' }
#' 
#' 
#' 
#' @source Hanretty, C. (2017). Areal interpolation and the UK’s referendum on EU
#' membership. _Journal of Elections, Public Opinion and Parties_, 27(4),
#' 466–483. \url{https://doi.org/10.1080/17457289.2017.1287081}
#' 
#' @source \url{https://docs.google.com/spreadsheets/d/1b71SDKPFbk-ktmUTXmDpUP5PT299qq24orEA0_TOpmw/edit#gid=579044181}
#' @source \url{http://www.bbc.co.uk/news/uk-northern-ireland-36616830}
"leave_votes_west"
